### ISSUE TITLE
It is difficult to get the width of the bars in various series

### ISSUE THREAD

**Issue by rjmcneill** [d3fc/d3fc#881]
[ohlcBase](https://github.com/ScottLogic/d3fc/blob/f752ac1cd19cae25566db4297ae7d5fec8150f53/src/series/ohlcBase.js#L21-L23) has a method for getting the width of the bars.

Could be useful to be able to get the bar width in:
- bar series
- errorBar
- boxPlot
- macd
- elder ray

It also looks like you can't change the bar width on:
- grouped bar series
- macd
- elder ray

**Comment by ColinEberhardt** [comment]
That's interesting, I don't think we'd ever considered that as being part of the public API. Out of interest, what are you using it for?

**Comment by rjmcneill** [comment]
We're looking to use it to decorate the crosshair to be the full width of the bars.
We're using it [here][1] on the primary chart at the minute, but are looking into it for the secondaries.

[1]: http://scottlogic.github.io/BitFlux/master/

**Comment by ColinEberhardt** [comment]
That makes sense. How about you split this into two PRs?

  1. Non-controversial, exposing bar width for the indicators you listed above - by using the rebind technique in your PR #882 
  2. Expose the `width` function for the other series types.

For (2) one thing to bear in mind is that the series first filter the data before computing the width. Therefore in order to get the correct width, users will also have to apply the filter, which would also have to be exposed (if it isn't already).

**Comment by rjmcneill** [comment]
Width seems to work ok for OHLC because the scales are bound and directly set.
For series that use multi however, it won't work as the scales are set after the multiseries is called.
The limitation is that for multiseries, you have to render once before calling the width function to get the correct value returned.

Couple of options:
- Have the simple implementation that needs to be rendered once. This works fine for bar, but for series like MACD, the scales get set after rendering, so this implementation won't work until after a render.
```JavaScript
    bar.width = function(data) {
        return calculateWidth(data);
    };

    function calculateWidth(data) {
        var filteredData = data.filter(base.defined);
        return barWidth(filteredData.map(crossAxisValueFunction()));
    }
```
- Ensure the scales and excessors are set before calling the width function.  
``` JavaScript
    macd.barWidth = function(data) {
        // Could also set the scales here too
        divergenceBar.xValue(xValue)
            .yValue(function(d, i) { return root(d).divergence; });
        return divergenceBar.width(data);
    };
```
- If the width isn't supposed to be part of the public API, leave it as is at the minute. I'll look into getting the width outside of d3fc by building up the functions externally.

**Comment by rjmcneill** [comment]
Managed to work around the problem by building up the necessary functions and data outside of d3fc.  It's a bit convoluted though.  The crosshair snapping issue we talked about is also demonstrated on this branch in the multi example.

[Branch with demos] (https://github.com/rjmcneill/d3fc/tree/demo/barWidth)
[Blob for width in macd] (https://github.com/rjmcneill/d3fc/blob/3b9c43528dc747fb1e79dc592fc544cfab43d189/visual-tests/robert-demos/macd.js#L43-L54)

If you run the visual tests, the demos are right at the top of the page.
Crosshair snapping problem is shown in the multi demo
